gematria/datasets/annotating_importer.cc

```diff
@@ -0,0 +1,340 @@
+// Copyright 2024 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gematria/datasets/annotating_importer.h"
+
+#include <cstdint>
+#include <string>
+#include <string_view>
+#include <tuple>
+#include <unordered_map>
+#include <vector>
+
+#include "absl/status/status.h"
+#include "absl/status/statusor.h"
+#include "absl/strings/numbers.h"
+#include "absl/strings/str_cat.h"
+#include "gematria/basic_block/basic_block_protos.h"
+#include "gematria/datasets/bhive_importer.h"
+#include "gematria/llvm/canonicalizer.h"
+#include "gematria/llvm/disassembler.h"
+#include "gematria/llvm/llvm_to_absl.h"
+#include "gematria/utils/string.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Object/Binary.h"
+#include "llvm/Object/ELFObjectFile.h"
+#include "llvm/Object/ELFTypes.h"
+#include "llvm/Object/ObjectFile.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Support/Error.h"
+#include "quipper/perf_parser.h"
+#include "quipper/perf_reader.h"
+
+namespace gematria {
+
+AnnotatingImporter::AnnotatingImporter(const Canonicalizer *canonicalizer)
+    : importer_(canonicalizer), perf_reader_(), perf_parser_(&perf_reader_) {
+  quipper::PerfParserOptions parser_opts;
+  parser_opts.do_remap = true;
+  parser_opts.discard_unused_events = true;
+  parser_opts.sort_events_by_time = false;
+  perf_parser_.set_options(parser_opts);
+}
+
+absl::Status AnnotatingImporter::LoadPerfData(std::string_view file_name) {
+  // Read and parse the `perf.data`-like file into something more tractable.
+  if (!perf_reader_.ReadFile(file_name.data())) {
+    return absl::InvalidArgumentError(
+        "The given `perf.data`-like file could not be read.");
+  }
+  if (!perf_parser_.ParseRawEvents()) {
+    return absl::InvalidArgumentError(
+        "The given `perf.data`-like file could not be parsed.");
+  }
+
+  return absl::OkStatus();
+}
+
+absl::StatusOr<std::vector<
+    std::tuple<uint64_t, uint64_t, std::vector<DisassembledInstruction>>>>
```

Comment:
As a general comment: most of the time, I'd prefer to avoid `std::pair` and `std::tuple` and use a struct instead. It often produces more readable code that needs less documentation.

In this specific case: do we need the start + end address? When `DisassembledInstruction` is filled in completely, then `start == instructions.front().address` and `end == instructions.back().address + instructions.back().machine_code.size()`, no?

Reply:
Yep, I realised this and had added a little TODO for myself just beneath on line 140 🙂 

```cpp
// TODO(vbshah): Consider removing the addresses and just using the
// `addr` fields on the instructions.
basic_blocks.emplace_back(begin_idx - file_offset,
                          end_idx - file_offset, instructions.value());
```